app/src/ui/discard-changes/discard-selection-dialog.tsx

```diff
@@ -50,13 +50,9 @@ export class DiscardSelection extends React.Component<
   IDiscardSelectionProps,
   IDiscardSelectionState
 > {
-  public constructor(props: IDiscardSelectionProps) {
-    super(props)
-
-    this.state = {
-      isDiscardingSelection: false,
-      confirmDiscardSelection: false,
-    }
+  public state: IDiscardSelectionState = {
```

Comment:
It feels really weird to me to declare state as public like this. Likely it's just a remnant of my c# days but given that this is a brand new pattern that we don't use anywhere else in the code base (that I know of) could we keep this PR focused on the flipping of the property and maybe discuss the pros/cons of switching patterns at our next sync?

Reply:
Oh yes! Didn't realize it's not used anywhere else in this codebase 😅  (I got confused because we're already using instance fields for other object fields like refs and instance methods for event handlers that need to be autobound, but you're right that the state it's currently always initialized inside the constructor).

For the upcoming discussion, a couple of points:
* Behind the scenes both forms are exactly equal. I agree that seeing the state being declared as `public` looks weird at first sight, but that's how it's actually typed in the React type declarations so this is just making it more obvious...
* I'm not a big fan of defining a constructor on React components: their usefulness is quite limited (they should only be used to initialize instance fields or instance methods - see the [React docs](https://reactjs.org/docs/react-component.html#constructor) -, which can be done via the instance field syntax). Any other logic should instead be included in the `componentDidMount()` method. Having components with an already defined constructor makes it easier to keep adding logic to it and end up with side effects there. I've found a few examples of components with side effects in the constructor in our codebase ([one of them](https://github.com/desktop/desktop/blob/development/app/src/ui/dialog/dialog.tsx#L158-L160)), so this is something that definitely ends up happening.
* As a con, in some components we currently have not-so-trivial-logic to calculate the state based on props ([example](https://github.com/desktop/desktop/blob/development/app/src/ui/add-repository/create-repository.tsx#L98-L119)). In these cases we would have to choose whether we keep the constructor or we create a `createInitialState()` method like we have in other components.